flood_adapt/object_model/hazard/event/historical.py

```diff
@@ -0,0 +1,373 @@
+    IEvent,
+    IEventModel,
+    Mode,
+)
+from flood_adapt.object_model.hazard.interface.forcing import (
+    ForcingSource,
+    ForcingType,
+)
+from flood_adapt.object_model.interface.scenarios import IScenario
+from flood_adapt.object_model.interface.site import ISite, Obs_pointModel
+from flood_adapt.object_model.io.unitfulvalue import UnitfulLength, UnitTypesLength
+
+
+def run_once_with_unique_args(func: Callable) -> Callable:
+    """Run the function only once for each unique combination of arguments, otherwise return the result from earlier immediately."""
+
+    def wrapper(self, *args: Tuple[Any], **kwargs: Dict[str, Any]) -> Any:
+        if not hasattr(self, "_called_args"):
+            self._called_args = {}
+
+        args_key = (
+            str(args) + str(sorted(kwargs.items())) if args or kwargs else "no_args"
+        )
+        if args_key in self._called_args.get(func.__name__, set()):
+            return  # Return immediately if the function has been called with these arguments
+
+        if func.__name__ not in self._called_args:
+            self._called_args[func.__name__] = set()
+
+        result = func(self, *args, **kwargs)
+        self._called_args[func.__name__].add(args_key)
+
+        return result
+
+    return wrapper
+
+
+class HistoricalEventModel(IEventModel):
+    """BaseModel describing the expected variables and data types for parameters of HistoricalNearshore that extend the parent class Event."""
+
+    ALLOWED_FORCINGS: dict[ForcingType, List[ForcingSource]] = {
+        ForcingType.RAINFALL: [ForcingSource.CONSTANT, ForcingSource.MODEL],
+        ForcingType.WIND: [ForcingSource.CONSTANT, ForcingSource.MODEL],
+        ForcingType.WATERLEVEL: [ForcingSource.CSV, ForcingSource.MODEL],
+        ForcingType.DISCHARGE: [ForcingSource.CONSTANT],
+    }
+
+
+class HistoricalEvent(IEvent):
+    MODEL_TYPE = HistoricalEventModel
+
+    attrs: HistoricalEventModel
+
+    def __init__(self):
+        self._site: ISite = db.Database().site
+        self._logger = FloodAdaptLogging().getLogger(__name__)
+
+    def process(self, scenario: IScenario):
+        """Preprocess, run and postprocess offshore model to obtain water levels for boundary condition of the overland model."""
+        self._scenario = scenario
+        self.meteo_ds = None
+        sim_path = self._get_simulation_path()
+
+        require_offshore_run = any(
+            forcing._source == ForcingSource.MODEL
+            for forcing in self.attrs.forcings.values()
+            if forcing is not None
+        )
+
+        if require_offshore_run:
+            self.download_meteo(self.attrs.time.start_time, self.attrs.time.end_time)
+            self.meteo_ds = self.read_meteo()
+
+            self._preprocess_sfincs_offshore(sim_path)
+            self._run_sfincs_offshore(sim_path)
+
+        if self.attrs.mode == Mode.risk:
+            self._process_risk_event()
+        else:
+            self._process_single_event(sim_path)
```

Comment:
if mode==risk it is an event set, which does not have a template (maybe we need to add one but it wouldn't be historic).